### PR TITLE
tests: handle case of `acceptance-test-config.yml` with no valid test cases

### PR DESCRIPTION
**Pull request by aaronsteers**
This PR handles the case of missing config in `acceptance-test-config.yml`. Specifically, it now runs successfully on `source-7shift`, which has the following definition:

```yml
# See [Connector Acceptance Tests](https://docs.airbyte.com/connector-development/testing-connectors/connector-acceptance-tests-reference)
# for more information about how to configure these tests
connector_image: airbyte/source-7shifts:dev
acceptance_tests:
  spec:
    tests:
      - spec_path: "manifest.yaml"
  connection:
    bypass_reason: "This is a builder contribution, and we do not have secrets at this time"
  discovery:
    bypass_reason: "This is a builder contribution, and we do not have secrets at this time"
  basic_read:
    bypass_reason: "This is a builder contribution, and we do not have secrets at this time"
  incremental:
    bypass_reason: "This is a builder contribution, and we do not have secrets at this time"
  full_refresh:
    bypass_reason: "This is a builder contribution, and we do not have secrets at this time"
```

This test config file is especially useless because `spec` is not the same as `manifest.yml` and so it's not really helpful at all in defining test cases.

Nevertheless, we test what we can and the tests succeed now:

image

## Summary by CodeRabbit

- **Tests**
  - Improved handling of missing test scenarios in acceptance test configuration, returning an empty list instead of raising an error when no tests are found for a category.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new standard test for validating connector specification output.
  - Added a property to easily access specification messages in test results.

- **Bug Fixes**
  - Improved handling of test scenarios to support cases where no scenario is provided.
  - Enhanced test scenario aggregation to include multiple categories and handle missing data more gracefully.
  - Expanded test job support to include the "spec" verb and optional test scenarios.

### CODE REVIEW

File: airbyte_cdk/test/standard_tests/_job_runner.py [hunk 1]
```diff
@@ -56,9 +56,9 @@ def spec(self, logger: logging.Logger) -> Any:
 
 def run_test_job(
     connector: IConnector | type[IConnector] | Callable[[], IConnector],
-    verb: Literal["read", "check", "discover"],
-    test_scenario: ConnectorTestScenario,
+    verb: Literal["spec", "read", "check", "discover"],
     *,
+    test_scenario: ConnectorTestScenario | None = None,
```
Comment (aaronsteers) [inline airbyte_cdk/test/standard_tests/_job_runner.py:61]: This now handles a "None" scenario - especially for the `spec` use case. If test_scenario is omitted, we presumably won't have a `config` but this is not needed for `spec` anyway, and there may be additional cases where it is not needed which we haven't considered as of yet.

File: airbyte_cdk/test/standard_tests/_job_runner.py [hunk 2]
```diff
@@ -81,9 +81,9 @@ def run_test_job(
         )
 
     args: list[str] = [verb]
-    if test_scenario.config_path:
+    if test_scenario and test_scenario.config_path:
         args += ["--config", str(test_scenario.config_path)]
-    elif test_scenario.config_dict:
+    elif test_scenario and test_scenario.config_dict:
```
Comment (aaronsteers) [inline airbyte_cdk/test/standard_tests/_job_runner.py:86]: Now we check if scenario exists before pulling 'config' from it.

File: airbyte_cdk/test/standard_tests/_job_runner.py [hunk 3]
```diff
@@ -112,12 +112,18 @@ def run_test_job(
...
+    if test_scenario and test_scenario.expect_exception:
+        # If the test scenario expects an exception, we need to set the
+        # `expect_exception` flag to True.
+        expect_exception = True
+
     result: entrypoint_wrapper.EntrypointOutput = entrypoint_wrapper._run_command(  # noqa: SLF001  # Non-public API
         source=connector_obj,  # type: ignore [arg-type]
         args=args,
-        expecting_exception=test_scenario.expect_exception,
+        expecting_exception=False if not test_scenario else expect_exception,
```
Comment (aaronsteers) [inline airbyte_cdk/test/standard_tests/_job_runner.py:124]: I forgot to update this after adding the dedicated variable. Fixes:

```suggestion
        expecting_exception=expect_exception,
```

File: airbyte_cdk/test/standard_tests/source_base.py [hunk 4]
```diff
@@ -64,6 +64,26 @@ def test_discover(
             test_scenario=scenario,
         )
 
+    def test_spec(
+        self,
+        # scenario: ConnectorTestScenario,  # No inputs needed for spec test
+    ) -> None:
+        """Standard test for `spec`."""
```
Comment (aaronsteers) [inline airbyte_cdk/test/standard_tests/source_base.py:71]:
```suggestion
    def test_spec(
        self,
        # scenario: ConnectorTestScenario,  # No inputs needed for spec test
    ) -> None:
        """Standard test for `spec`.

        This test does not require a `scenario` input, since `spec`
        does not require any inputs.
        
        We assume `spec` should always succeed and it should always generate 
        a valid `SPEC` message.

        Note: the parsing of messages by type also implicitly validates that
        the generated `SPEC` message is valid JSON.
        """
```

File: airbyte_cdk/test/standard_tests/declarative_sources.py [hunk 5]
```diff
@@ -64,7 +64,7 @@ def components_py_path(cls) -> Path | None:
     @classmethod
     def create_connector(
         cls,
-        scenario: ConnectorTestScenario,
+        scenario: ConnectorTestScenario | None,
```
Comment (aaronsteers) [inline airbyte_cdk/test/standard_tests/declarative_sources.py:67]: The scenario can now accept `None` but it does not default to `None`. For now at least, this prevents us from forgetting to provide `scenario`, while still allowing us to do so explicitly when needed.

File: airbyte_cdk/test/standard_tests/_job_runner.py [hunk 6]
```diff
@@ -112,12 +112,18 @@ def run_test_job(
     # This is a bit of a hack because the source needs the catalog early.
     # Because it *also* can fail, we have to redundantly wrap it in a try/except block.
 
+    expect_exception = False
+    if test_scenario and test_scenario.expect_exception:
```
Comment (dbgold17) [inline airbyte_cdk/test/standard_tests/_job_runner.py:116]: could probably restructure this logic a bit so you aren't repeatedly checking if the scenario isn't `None`

File: airbyte_cdk/test/standard_tests/connector_base.py [hunk 7]
```diff
@@ -89,7 +89,7 @@ def get_test_class_dir(cls) -> Path:
     @classmethod
     def create_connector(
         cls,
-        scenario: ConnectorTestScenario,
+        scenario: ConnectorTestScenario | None,
```
Comment (dbgold17) [inline airbyte_cdk/test/standard_tests/connector_base.py:92]: why do we need a scenario at all for this method?

File: airbyte_cdk/test/standard_tests/source_base.py [hunk 8]
```diff
@@ -64,6 +64,33 @@ def test_discover(
...
+
+        Note: the parsing of messages by type also implicitly validates that
+        the generated `SPEC` message is valid JSON.
+        """
+        result = run_test_job(
+            verb="spec",
+            test_scenario=None,
+            connector=self.create_connector(scenario=None),
+        )
+        if result.errors:
```
Comment (dbgold17) [inline airbyte_cdk/test/standard_tests/source_base.py:84]: This is outside the scope of this PR, but I think I need to better understand the structure of the standard test code. At first glance it seems somewhat duplicative to be checking for errors here but also have some logic in the `run_test_job` method to check for whether exceptions were thrown

**Review by dbgold17 — APPROVED** [review]
approving with some comments

File: airbyte_cdk/test/standard_tests/connector_base.py (hunk 7 above)
Comment (aaronsteers) [reply airbyte_cdk/test/standard_tests/connector_base.py:92]: Great question. In the long run, I would like to not need it. But for legacy reasons, many (_most?_) connectors cannot properly initialize themselves without having the config, catalog, and/or state inputs. Ideally, these inputs would only be passed when invoking the verbs (`read()`, `discover()`, etc.) but for now, they are part of the class constructor - and connectors (esp. S3, which I have experience with) may act unpredictably if the inputs are not passed to the constructor.

I have another thread on this topic from earlier today with @maxi297 here:

- https://github.com/airbytehq/airbyte-python-cdk/pull/527#discussion_r2075956954

File: airbyte_cdk/test/standard_tests/_job_runner.py (hunk 6 above)
Comment (aaronsteers) [reply airbyte_cdk/test/standard_tests/_job_runner.py]: @dbgold17 - Per your feedback, I've updated the implementation so that scenarios can now have an empty config. This reduces the not-null checks we otherwise would have to do in this implementation.

File: airbyte_cdk/test/standard_tests/source_base.py (hunk 8 above)
Comment (aaronsteers) [reply airbyte_cdk/test/standard_tests/source_base.py]: You are right to raise it. In fact, as long as the `run_test_job` doesn't get a scenario that expects failure, it will raise the error and message about it for us. So, I've updated the implementation to simply have a code-comment that the hard-fail scenario is already handled. Then, all that is left to do is check that we have 1 and exactly 1 `SPEC` output.

Here is the part in `run_test_job` which does the check already:

- https://github.com/airbytehq/airbyte-python-cdk/blob/72e73ecc3e04aae6513c47ef5f6f56675266b803/airbyte_cdk/test/standard_tests/_job_runner.py#L123-L127

File: airbyte_cdk/test/standard_tests/models/scenario.py [hunk 9]
```diff
@@ -13,6 +13,7 @@
 from typing import Any, Literal, cast
 
 import yaml
+from numpy import empty
```
Comment (aaronsteers) [inline airbyte_cdk/test/standard_tests/models/scenario.py:16]:
```suggestion
```
